networkx/algorithms/tsp.py

```diff
@@ -0,0 +1,319 @@
+    >>> G = nx.DiGraph()
+    >>> G.add_weighted_edges_from({('A', 'B', 3),
+    ...                           ('A', 'C', 17), ('A', 'D', 14), ('B', 'A', 3),
+    ...                          ('B', 'C', 12), ('B', 'D', 16), ('C', 'A', 13),
+    ...                           ('C', 'B', 12), ('C', 'D', 4), ('D', 'A', 14),
+    ...                           ('D', 'B', 15), ('D', 'C', 2)})
+    >>> sol = nx.simulated_annealing_tsp(G, 'D')
+    >>> sol[0]
+    ['D', 'C', 'B', 'A', 'D']
+    >>> sol[1]
+    31.0
+    >>> sol = nx.simulated_annealing_tsp(G, 'D', sol=['D', 'B', 'A', 'C', 'D'])
+    >>> sol[0]
+    ['D', 'C', 'B', 'A', 'D']
+    >>> sol[1]
+    31.0
+
+    Notes
+    -----
+    Simulated Annealing is a metaheuristic local search algorithm.
+    The main characteristic of this algorithm is that it accepts
+    even solutions which lead to the increase of the cost in order
+    to escape from low quality local optimal solutions.
+
+    This algorithm needs an initial solution. This solution can be
+    constructed by a simple greedy algorithm. At every iteration, it
+    selects thoughtfully a neighbor solution.
+    Consider c(x) cost of current solution and c(x') cost of
+    neighbor solution.
+    If c(x') - c(x) <= 0 then neighbor solution becomes current
+    solution for the next iteration. Otherwise, algorithm accepts
+    neighbor solution to become current solution for the next iteration
+    with probability p = exp - ([c(x') - c(x)] / temp).
+
+    Temp is parameter of algorithm and represents temperature in every
+    iteration.
+
+    For more information and how algorithm is inspired see:
+    http://en.wikipedia.org/wiki/Simulated_annealing
+    """
+    if sol is None:
+        # Construct an initial solution using a greedy algorithm.
+        sol, cost = greedy_tsp(G, source, weight=weight)
+
+    else:
+        # Calculate the cost of initial solution and make the essential checks for graph.
+        if not _is_completed(G):
+            raise nx.NetworkXError('Given graph is not completed.')
+        if not nx.is_weighted(G, weight=weight):
+            raise nx.NetworkXError('Given graph is not weighted.')
+        cost = _calculate_cost(G, sol, weight=weight)
+
+    count = 0
+    best_sol = list(sol)
+    best_cost = cost
+    while count <= outer_iter and temp > 0:
+        count += 1
+        for i in range(0, inner_iter):
+            adj_sol = _apply_move(sol, move)
+            adj_cost = _calculate_cost(G, adj_sol, weight=weight)
+            delta = adj_cost - cost
+            if delta <= 0:
+
+                # Set current solution the adjacent solution.
+                sol = list(adj_sol)
+                cost = adj_cost
+
+                if cost < best_cost:
+                    count = 0
+                    best_sol = list(sol)
+                    best_cost = cost
+            else:
+
+                # Accept even a worse solution with probability p.
+                p = math.exp(- (delta / temp))
```

Comment:
Add `from __future__ import division` to head of file to ensure true division occurs on Python 2.